File: internal/database/models/task.go
package models

import (
	"database/sql"
	"time"

	"gorm.io/datatypes"
	"gorm.io/gorm"
)

// TaskType is the type used to describe the type of task
type TaskType string

const (
	// TaskTypeShell is a task for starting external process via shell
	TaskTypeShell TaskType = "SHELL"
	// TaskTypeWebhook is a task type that triggers an external webhook
	TaskTypeWebhook TaskType = "WEBHOOK"
)

// Task define the data structure of a timed task
type Task struct {
	// gorm base model
	gorm.Model

	// Name of task
	Name string

	// Type indicates the type of this task
	Type TaskType

	// Configuration
	Configuration datatypes.JSON

	// Rules of task
	Rules []TaskRule

	// LastRunningAt indicates the last running time of the current task
	LastRunningAt sql.NullTime

	// LastRunningTime indicates the time consumed for the last run of the current task
	LastRunningTime int64

	// Status indicates the current task status
	Status Status
}

// TaskConfigurationShell define the data structure of shell task configuration
type TaskConfigurationShell struct {
	Command string        `json:"command"`
	Timeout time.Duration `json:"timeout"`
}

// IsTaskConfiguration is a flag function for task configuration
func (tcs TaskConfigurationShell) IsTaskConfiguration() {}

// TaskConfigurationWebhook define the data structure of webhook task configuration
type TaskConfigurationWebhook struct {
	URL    string `json:"url"`
	Method string `json:"method"`
}

// IsTaskConfiguration is a flag function for task configuration
func (tcw TaskConfigurationWebhook) IsTaskConfiguration() {}

// Implement GraphQL's models.Model interface to simplify data assembly,
// the data corresponding to models.Model is gorm.Model

// IsModel is a flag field of the GraphQL Model interface
func (t Task) IsModel() {}

// GetID is a function to get the field value of GraphQL Model
func (t Task) GetID() int64 {
	return int64(t.ID)
}

// GetCreatedAt is a function to get the field value of GraphQL Model
func (t Task) GetCreatedAt() int64 {
	return t.CreatedAt.UnixMilli()
}

// GetUpdatedAt is a function to get the field value of GraphQL Model
func (t Task) GetUpdatedAt() int64 {
	return t.UpdatedAt.UnixMilli()
}

File: src/database/entities/task.ts
import { Collection, Entity, Enum, OneToMany, Property } from "@mikro-orm/core";
import { BaseEntity, Status } from "@/database/entities/common";
import { TaskRule } from "@/database/entities/task_rule";
import { Job } from "@/database/entities/job";
import {
  Task as GQLTask,
  HttpMethod as GQLHttpMethod,
} from "@/generated/graphql";

export enum TaskType {
  Shell = "SHELL",
  Webhook = "WEBHOOK",
}

export interface TaskConfigurationShell {
  command: string;
  timeout: number;
}

export interface TaskConfigurationWebhook {
  url: string;
  method: GQLHttpMethod;
}

export type TaskConfiguration =
  | TaskConfigurationShell
  | TaskConfigurationWebhook;

@Entity({ tableName: "tasks" })
export class Task extends BaseEntity {
  constructor(
    name: string,
    type: TaskType,
    configuration: TaskConfiguration,
    status?: Status
  ) {
    super();
    this.name = name;
    this.type = type;
    this.configuration = configuration;
    if (status) this.status = status;
  }

  @Property()
  name: string;

  @Enum(() => TaskType)
  type: TaskType;

  @Property({ type: "json" })
  configuration: TaskConfiguration;

  @OneToMany(() => TaskRule, (rule) => rule.task)
  rules = new Collection<TaskRule>(this);

  @OneToMany(() => Job, (job) => job.task)
  jobs = new Collection<Job>(this);

  @Property({ fieldName: "last_running_at", nullable: true })
  lastRunningAt?: Date;

  @Property({ fieldName: "last_running_time", nullable: true })
  lastRunningTime?: number;

  @Enum({ items: () => Status, type: "Status" })
  status: Status = Status.Enabled;

  toGQLType(): GQLTask {
    return {
      id: this.id,
      createdAt: this.createdAt,
      updatedAt: this.updatedAt,
      name: this.name,
      type: this.type,
      configuration: this.configuration,
      lastRunningAt: this.lastRunningAt,
      lastRunningTime: this.lastRunningTime,
      jobs: [],
      rules: [],
      status: this.status,
    };
  }
}
